nym-vpn-core/crates/nym-vpn-lib/src/bandwidth_controller.rs

```diff
@@ -389,77 +532,186 @@ impl BandwidthController {
         Ok(remaining_bandwidth)
     }
 
-    async fn check_bandwidth(&mut self, entry: bool, current_period: Duration) -> Option<Duration> {
-        let (wg_metadata_client, current_depletion_rate) = if entry {
-            (
-                &mut self.wg_entry_gateway_client,
-                &mut self.entry_depletion_rate,
-            )
+    pub(crate) async fn request_upgrade_mode_recheck(
+        &mut self,
+        entry: bool,
+    ) -> Result<bool, SpecificGatewayError> {
+        let bw_client = if entry {
+            &mut self.wg_entry_gateway_client
         } else {
-            (
-                &mut self.wg_exit_gateway_client,
-                &mut self.exit_depletion_rate,
-            )
+            &mut self.wg_exit_gateway_client
+        };
+
+        let Some(upgrade_mode_jwt) = self
+            .ticket_provider
+            .get_upgrade_mode_token()
+            .await
+            .map_err(|source| SpecificGatewayError::UpgradeModeTokenRequest {
```

Comment:
If I'm reading this right, it's more of a storage error than a `SpecificGatewayError` one, because it would happen regardless of the gateway used, unlike a top up or a bandwidth query right?

Reply:
Yes, I agree, this error layer is supposed to be used for differentiating between entry and exit specific errors